tests/test_distance.py

```diff
@@ -1,13 +1,98 @@
 import pytest
-
+import geopandas as gpd
 from gpbp.distance import population_served
 from gpbp.layers import AdmArea
 
+@pytest.fixture
+def population_data():
+    # This population data is made up but located around Dili, Timor-Leste
+    pop_gdf = pd.DataFrame(
+        {
+            "ID": [0, 1, 2],
+            "longitude" : [125.5, 125.7, 124.8],
+            "latitude" : [-8.6, -8.7, -9.1],
+            "population": [0.1, 0.2, 0.05],
+        }
+    )
+    pop_gdf = gpd.GeoDataFrame(pop_gdf, geometry=gpd.points_from_xy(pop_gdf.longitude, pop_gdf.latitude))
+    return pop_gdf
+
+@pytest.fixture
+def facilities_data():
+    # These are the Timor-Leste Hospitals with IDs 8 and 19
+    fac_gdf = pd.DataFrame(
+        {
+            "ID": [0, 1],
+            "osmid": [1258236683.0, 1150722282.0],
+            "longitude" : [125.52661711024034, 126.99755338864472],
+            "latitude" : [-8.557962051100692, -8.5238558567869],
+            "geometry": [None, None],
+            # Geometry is dropped in the first step of population_served anyways and never used again (except ID column).
+        }
+    )
+    return fac_gdf
+
+@pytest.fixture
+def road_network():
+    # I need the road network and cannot just hand-make that data.
+    return pickle.load(open("tests/test_assets/road_network_TLS.pkl", "rb"))
+
+def test_population_served_invalid_strategy(population_data, facilities_data):
+    with pytest.raises(Exception) as exc_info:
+        population_served(
+            pop_gdf=population_data,
+            fac_gdf=facilities_data,
+            data_as_key="facilities",
+            distance_type="length",
+            distance_values=[1000, 2000],
+            route_mode="drive",
+            strategy="XXX",
+            access_token=None,
+            road_network=None,
+        )
+
+    assert "Invalid strategy" in str(exc_info.value)
+
+# This parameterization works, but would need to provide a valid access token for mapbox.
+# Instead, I guess we should mock the API call to mapbox and return some pre-saved data.
+# Also: the number of different parameters means the tests take now 50 seconds to run (on my machine), about 10x longer than with only one case.
+@pytest.mark.parametrize(
+        "distance_type, distance_values, strategy, access_token",
+        [
+            ("length", [0.1, 1000], "osm", None),
+            ("travel_time", [10], "osm", None),
+            # ("mapbox", None),
+        ]
+)
+def test_population_served(population_data, facilities_data, road_network, distance_type, distance_values, strategy, access_token):
+    serve_df = population_served(
+        pop_gdf=population_data,
+        fac_gdf=facilities_data,
+        data_as_key="facilities",
+        distance_type=distance_type,
+        distance_values=distance_values,
+        route_mode="drive", # route_mode is irrelevant for strategy osm
+        strategy=strategy,
+        access_token=access_token,
+        road_network=road_network,
+    )
```

Comment:
I think this function is quite complex, we might want to refactor it into smaller ones. It might be one of those cases that Catalina mentioned where it's worth splitting the big function into smaller ones before doing the testing.

But, for now, have you considered creating a mock `dist_dict` so that we would not have to run `calculate_isopolygons_Mapbox` nor `calculate_isopolygons_graph` when testing this function? We could do the testing of those two functions separately.

Reply:
Good idea to create the mock